src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -720,16 +717,244 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
+
+                Debug.Assert(countOfElements == arguments.Count);
+                var elementLocationsBuilder = ArrayBuilder<Location>.GetInstance(arguments.Count);
+
+                for (int i = 0, l = arguments.Count; i < l; i++)
+                {
+                    ArgumentSyntax argumentSyntax = arguments[i];
+                    IdentifierNameSyntax nameSyntax = argumentSyntax.NameColon?.Name;
+
+                    if (nameSyntax != null)
+                    {
+                        elementLocationsBuilder.Add(nameSyntax.Location);
+                    }
+                    else
+                    {
+                        elementLocationsBuilder.Add(argumentSyntax.Location);
+                    }
+                }
+
+                elementlocations = elementLocationsBuilder.ToImmutableAndFree();
+            }
+
+            var declaredTupleType = TupleTypeSymbol.Create(typeLocation, inferredElementTypes.ToImmutableAndFree(),
+                                                           elementlocations, tupleLiteral.ArgumentNamesOpt,
+                                                           this.Compilation, ref useSiteDiagnostics);
+
+            return tupleLiteral.Update(elements.ToImmutableAndFree(), tupleLiteral.ArgumentNamesOpt, declaredTupleType);
+        }
+
+        /// <summary>
+        /// Reclassify tuple literal to its natural type.
+        /// 
+        /// A literal that has a natural type should not add any errors to <paramref name="diagnostics"/>
+        /// during reclassification.
+        /// </summary>
+        private BoundConversion ReclassifyTupleLiteral(BoundTupleLiteral tupleLiteral, DiagnosticBag diagnostics, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            tupleLiteral = AdjustTupleLiteralBasedOnNaturalTypeInference(tupleLiteral, diagnostics, ref useSiteDiagnostics);
+            Debug.Assert((object)tupleLiteral.DeclaredOrTargetTupleTypeOpt != null);
 
-            if (hasNaturalType)
+            return new BoundConversion(tupleLiteral.Syntax, tupleLiteral, Conversion.ImplicitTupleLiteral,
+                                       @checked: false, explicitCastInCode: false, constantValueOpt: null,
+                                       type: tupleLiteral.DeclaredOrTargetTupleTypeOpt);
+        }
+
+        private BoundTupleLiteral ReclassifyTupleLiteralInErrorRecoveryMode(BoundTupleLiteral tupleLiteral)
+        {
+            var diagnostics = DiagnosticBag.GetInstance();
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var result = AdjustTupleLiteralBasedOnNaturalTypeInference(tupleLiteral, diagnostics, ref useSiteDiagnostics);
+            diagnostics.Free();
+            return result;
+        }
+
+        /// <summary>
+        /// Does tuple literal have natural type?
+        /// </summary>
+        private static bool TupleLiteralHasNaturalType(BoundTupleLiteral tupleLiteral)
+        {
```

Comment:
This would also be much cheaper if tuple literal remembered its natural type right after binding.